### ISSUE TITLE
Allow ExUnit to clean up more than just the `tag` module attribute

### ISSUE THREAD

**Issue by bcardarella** [elixir-lang/elixir#4232]
I'd like to propose that the `test` macro can be instructed to delete more than just the `@tag` attribute.

Use case:

In EctoFixtures I'd like to have the following API:

```elixir
@fixtures :foo
test "some test", %{data: data} do
  ...
end
```

I *could* use the `@tag` attribute:

```elixir
@tag fixtures: :foo
test "some test", %{data: data} do
  ...
end
```

but the problem comes when someone wants to use multiple fixtures:

```elixir
@tag fixtures :foo
@tag fixtures :bar
```

The attribute accumulator will just overwrite `:fixtures` on the `context` object passed into the test. I could ask that an array is passed:

```elixir
@tag fixtures: [:foo, :bar]
```

But now the issue is when I want to support more options being passed in, it becomes messy. Instead it would be nice if I could write:

```elixir
@fixures :foo, insert: false
@fixtures :bar
```

I believe that ExUnit would have to be instructed on which module attributes to delete after each `test` macro: https://github.com/elixir-lang/elixir/blob/master/lib/ex_unit/lib/ex_unit/case.ex#L317

I'm happy to implement and PR.

**Comment by josevalim** [comment]
@bcardarella What would add meaning to the `@fixtures` module attribute?

**Comment by josevalim** [comment]
I mean, we could erase it, but usually in order to make use of it you would have to fetch it explicitly anyway and you could use this opportunity to delete it.

**Comment by bcardarella** [comment]
@josevalim perhaps I am missing where I can do that. From what I've seen fetching the module attribute during the `setup` will have after compilation, which means that all `@fixture` declarations will have been accumulated. Where I only care about the ones directly associated with that test, similar to how `@tag` is being used.

**Comment by josevalim** [comment]
Exactly. We do this because we have the `test` macro and we call some particular code. So the only way you can make use of `@fixtures` is if you also have a `test` macro, that reads `@fixtures` and call today's `test` macro. But if you do that, you can clean it up yourself.

To be more explicit: do you have a working version of your @fixtures today? It know it can work only for a single test for now, because we don't clean up, but if you have an example it will help me see how it is being used.

**Comment by bcardarella** [comment]
@josevalim I'll add a branch in a bit with a demo for you to look at

**Comment by bcardarella** [comment]
@josevalim 

Here is the implementation: https://github.com/dockyard/ecto_fixtures/blob/bc-fixture-tag-example/lib/ecto/fixtures.ex#L25-L46

and how it would be used (assuming a single test, not possible with multiple tests with above implementation) https://github.com/dockyard/ecto_fixtures/blob/bc-fixture-tag-example/test/module_attributes_test.exs#L10-L15

**Comment by josevalim** [comment]
Thank you! the implementation above wouldn't work after changing ExUnit either, because on before_compile you would have only the fixtures for the last test. The only way to make this work is by replacing ExUnit.Case.test by your own test macro that reads the attribute and calls ExUnit.Case.test. Once that is done, you won't need to change ExUnit as well because you can take care of the cleaning up too.

**Comment by bcardarella** [comment]
@josevalim what I am suggesting in this issue is a modification of how ExUnit.Case works. Where an additional number of module attributes can used along side `@tag` in ExUnit.Case. If you'd like I can PR that for discussion

**Comment by bcardarella** [comment]
I'm hesitant to go the route of wrapping ExUnit.Case in my library as it may not be compatible with other library that decide to do the same.

**Comment by josevalim** [comment]
@bcardarella let's try a PR then :) I can't guarantee it will be merged but I think it will progress this discussion!

**Comment by bcardarella** [comment]
Understood, I know its a long shot but let's see :)

**Comment by josevalim** [comment]
Closing in favor of the PR. :D
